src/modules/connections/services/ConnectionService.ts

```diff
@@ -418,17 +446,88 @@ export class ConnectionService {
     return this.connectionRepository.getSingleByQuery({ threadId })
   }
 
+  private async getRouting(mediationRecord: MediationRecord | undefined, routingKeys: string[], my_endpoint?: string) {
+    let endpoint
+    if (mediationRecord) {
+      routingKeys = [...routingKeys, ...mediationRecord.routingKeys]
+      endpoint = mediationRecord.endpoint
+    }
+    // Create and store new key
+    const did_data = await this.wallet.createDid()
+    if (mediationRecord) {
+      // new did has been created and mediator needs to be updated with the public key.
+      mediationRecord = await this.keylistUpdatdAndAwait(mediationRecord, did_data[1])
+    } else {
+      // TODO: register recipient keys for relay
+      // TODO: check that recipient keys are in wallet
+    }
+    endpoint = endpoint ?? my_endpoint ?? this.config.getEndpoint()
+    const result = { mediationRecord, endpoint, routingKeys, did: did_data[0], verkey: did_data[1] }
+    return result
+  }
+
+  public async keylistUpdatdAndAwait(
+    mediationRecord: MediationRecord,
+    verKey: string,
+    timeout = 15000 // TODO: this should be a configurable value in agent config
+  ): Promise<MediationRecord> {
+    const message = this.createKeylistUpdateMessage(verKey)
+    const connection = await this.getById(mediationRecord.connectionId)
+
+    const sendUpdateKeylist = async () => {
+      message.setReturnRouting(ReturnRouteTypes.all) // return message on request response
+      const outboundMessage = createOutboundMessage(connection, message)
+      await this.messageSender.sendMessage(outboundMessage)
+    }
+    const condition = async (event: KeylistUpdatedEvent) => {
+      return mediationRecord.id === event.payload.mediationRecord.id
+    }
+    const results = await waitForEvent(
+      sendUpdateKeylist,
+      RoutingEventTypes.RecipientKeylistUpdated,
+      condition,
+      timeout,
+      this.eventEmitter
+    )
+    return (results as KeylistUpdatedEvent).payload.mediationRecord
+  }
```

Comment:
I think this is a good case for observables. They take away the complexities of events + timing + conditions etc... I rewrote the method using rxjs and removed the `waitForEvent` call. Only addition is the following helper method to `EventEmitter.ts`. WDYT?


Event Emitter:
```ts
  // Add this method to EventEmitter.ts
  public observable<T extends BaseEvent>(event: T['type']) {
    return new Observable<T>((subscriber) => {
      const listener = (event: T) => {
        subscriber.next(event)
      }

      this.on<T>(event, listener)

      return () => {
        this.off<T>(event, listener)
      }
    })
  }
```

Keylist update event. The methods inside the `pipe` method are imported from `rxjs/operators`
```ts

  public async keylistUpdateAndAwait(
    mediationRecord: MediationRecord,
    verKey: string,
    timeoutMs = 15000 // TODO: this should be a configurable value in agent config
  ): Promise<MediationRecord> {
    const message = this.createKeylistUpdateMessage(verKey)
    const connection = await this.getById(mediationRecord.connectionId)

    // Create observable for event
    const observable = this.eventEmitter.observable<KeylistUpdatedEvent>(RoutingEventTypes.RecipientKeylistUpdated)

    // Apply required filters to observable stream and create promise to subscribe to observable
    const keylistUpdatePromise = observable
      .pipe(
        // Only take event for current mediation record
        filter((event) => mediationRecord.id === event.payload.mediationRecord.id),
        // Only wait for first event that matches the criteria
        first(),
        // Do not wait for longer than specified timeout
        timeout(timeoutMs)
      )
      .toPromise()

    message.setReturnRouting(ReturnRouteTypes.all) // return message on request response
    const outboundMessage = createOutboundMessage(connection, message)
    await this.messageSender.sendMessage(outboundMessage)

	// Await the observable promise
    const keylistUpdate = await keylistUpdatePromise
    return keylistUpdate.payload.mediationRecord
  }
```

Reply:
updated.